### PR TITLE
ci(benchmarks): ensure warmup iterations and data dumps

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

Ensure that the benchmarks run are preceded by warmup iterations to eliminate transient effects in the measures. Also dump the data to file to be collected for analysis

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by nizox — DISMISSED** [review]

File: riotfile.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
             name="benchmarks",
             pys=select_pys(),
             pkgs={"pytest-benchmark": latest, "msgpack": latest},
-            command="pytest {cmdargs} tests/benchmarks",
+            command="pytest --benchmark-warmup=on --benchmark-autosave {cmdargs} tests/benchmarks",
```
Comment (brettlangdon) [inline riotfile.py:157]: do we want autosave?

will we just overwrite with each serial run of `pytest`? (this is the problem we have with junit.xml files, see `conftest.py`)

Comment (P403n1x87) [reply riotfile.py:157]: The bigger picture is to collect that data along with whatever else we produce in CI and upload to an s3 bucket. File names will include commit hashes so there shouldn't be collisions

Comment (brettlangdon) [reply riotfile.py:157]: If you do `riot run benchmark`, then running with py 3.7 will overwrite the file generated for py 3.6. so after the run you'll only have 1 file, when you should have 1 per-python version. 

this is why in `conftest.py` we dynamically change the name of the `junit.xml` to `junit.<pid>.xml` (probably should be a uuid in hindsight, but seems fine for now)

Comment (P403n1x87) [reply riotfile.py:157]: Ah right, of course!

**Comment by P403n1x87** [comment]
The major source of variance turns out to be caused by the garbage collector

**Review by Kyle-Verhoog — APPROVED** [review]

File: riotfile.py (hunk 1 above)
Comment (P403n1x87) [reply riotfile.py:171]: @brettlangdon the default name had a timestamp so slim chance of collision, but I went for a custom name that includes a timestamp and the interpreter version info.

Comment (brettlangdon) [reply riotfile.py]: oh cool, yeah, timestamp probably would have been enough, but 👍🏻 to including interpreter name!

File: conftest.py [hunk 2]
```diff
@@ -31,6 +33,10 @@ def pytest_configure(config):
         # DEV: `ext` will contain the `.`, e.g. `.xml`
         config.option.xmlpath = "{0}.{1}{2}".format(fname, os.getpid(), ext)
 
+    # Save per-interpreter benchmark results.
+    if config.pluginmanager.hasplugin("benchmark"):
```
Comment (brettlangdon) [inline conftest.py:37]: we want this only for `CI=true`?

**Review by brettlangdon — APPROVED** [review]

File: conftest.py (hunk 2 above)
Comment (P403n1x87) [reply conftest.py:37]: I'd thought so. Locally one might just look at the more readable pytest output.

Comment (brettlangdon) [reply conftest.py:37]: 👍🏻 I sort of figured the same for `junit.xml`

**Comment by majorgreys** [comment]
> The major source of variance turns out to be caused by the garbage collector

Does this mean that enabling warmups in this PR does not help reduce variance or that variance does not change at all?

**Comment by P403n1x87** [comment]
> > The major source of variance turns out to be caused by the garbage collector
> 
> Does this mean that enabling warmups in this PR does not help reduce variance or that variance does not change at all?

There doesn't seem to be a significant difference between the two cases, but I think it's still good practice to allow for a warm-up step.

**Review by Yun-Kim — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**